L1Trigger/L1TTrackMatch/src/L1TkMuMantra.cc

```diff
@@ -0,0 +1,214 @@
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include "L1Trigger/L1TTrackMatch/interface/L1TkMuMantra.h"
+#include "TH2.h"
+
+using namespace L1TkMuMantraDF;
+
+L1TkMuMantra::L1TkMuMantra(std::vector<double>& bounds, TFile* fIn_theta, TFile* fIn_phi, std::string name = "mantra")
+    : wdws_theta_(bounds.size() - 1, MuMatchWindow()), wdws_phi_(bounds.size() - 1, MuMatchWindow()) {
+  name_ = name;
+
+  safety_factor_l_ = 0.0;
+  safety_factor_h_ = 0.0;
+
+  sort_type_ = kMaxPt;
+
+  // copy boundaries
+  nbins_ = bounds.size() - 1;
+  for (double b : bounds)
+    bounds_.push_back(b);
+
+  // now load in memory the TF1 fits
+
+  for (int ib = 0; ib < nbins_; ++ib) {
+    std::string wdn;
+    std::string nml;
+    std::string nmc;
+    std::string nmh;
+    TF1* fl;
+    TF1* fc;
+    TF1* fh;
+
+    wdn = name_ + std::string("_wdw_theta_") + std::to_string(ib + 1);
+    nml = std::string("fit_low_") + std::to_string(ib + 1);
+    nmc = std::string("fit_cent_") + std::to_string(ib + 1);
+    nmh = std::string("fit_high_") + std::to_string(ib + 1);
+
+    fl = (TF1*)fIn_theta->Get(nml.c_str());
+    fc = (TF1*)fIn_theta->Get(nmc.c_str());
+    fh = (TF1*)fIn_theta->Get(nmh.c_str());
+    if (fl == nullptr || fc == nullptr || fh == nullptr) {
+      if (verbosity_ > 0) {
+        LogTrace("L1TkMuMantra") << "... fit theta low  : " << fl << std::endl;
+        LogTrace("L1TkMuMantra") << "... fit theta cent : " << fc << std::endl;
+        LogTrace("L1TkMuMantra") << "... fit theta high : " << fh << std::endl;
+      }
+      throw std::runtime_error("L1TkMuMantra : Could not init theta");
+    }
+    wdws_theta_.at(ib).SetName(wdn);
+    wdws_theta_.at(ib).SetLower(fl);
+    wdws_theta_.at(ib).SetCentral(fc);
+    wdws_theta_.at(ib).SetUpper(fh);
+
+    wdn = name_ + std::string("_wdw_phi_") + std::to_string(ib + 1);
+    nml = std::string("fit_low_") + std::to_string(ib + 1);
+    nmc = std::string("fit_cent_") + std::to_string(ib + 1);
+    nmh = std::string("fit_high_") + std::to_string(ib + 1);
+    fl = (TF1*)fIn_phi->Get(nml.c_str());
+    fc = (TF1*)fIn_phi->Get(nmc.c_str());
+    fh = (TF1*)fIn_phi->Get(nmh.c_str());
+    if (fl == nullptr || fc == nullptr || fh == nullptr) {
+      if (verbosity_ > 0) {
+        LogTrace("L1TkMuMantra") << "... fit phi low  : " << fl << std::endl;
+        LogTrace("L1TkMuMantra") << "... fit phi cent : " << fc << std::endl;
+        LogTrace("L1TkMuMantra") << "... fit phi high : " << fh << std::endl;
+      }
+      throw std::runtime_error("L1TkMuMantra : Could not init phi");
+    }
+    wdws_phi_.at(ib).SetName(wdn);
+    wdws_phi_.at(ib).SetLower(fl);
+    wdws_phi_.at(ib).SetCentral(fc);
+    wdws_phi_.at(ib).SetUpper(fh);
+  }
+}
+
+int L1TkMuMantra::findBin(double val) {
+  // FIXME: not the most efficient, nor the most elegant implementation for now
```

Comment:
if the bounds are sorted, this could use binary search

Reply:
ok.  can do in the next version when optimizing the computing performance of the algos.